File: python_sdk/src/reality_capture/specifications/import_point_cloud.py
from pydantic import BaseModel, Field
from typing import Optional
from enum import Enum
from reality_capture.specifications.geometry import BoundingBox, Point3d


class ImportPCInputs(BaseModel):
    scene: str = Field(description="Reality data id of ContextScene to process")


class ImportPCOutputs(BaseModel):
    scan_collection: str = Field(description="Output reality data id for scan collection", alias="scanCollection")
    scene: Optional[str] = Field(default=None,
                                 description="Output reality data id for context scene referencing scan collection")


class ImportPCSpecifications(BaseModel):
    inputs: ImportPCInputs = Field(description="Inputs for importing a point cloud")
    outputs: ImportPCOutputs = Field(description="Outputs for point cloud import")


class ImportPCOutputsCreate(Enum):
    SCAN_COLLECTION = "scanCollection"
    SCENE = "scene"


class ImportPCSpecificationsCreate(BaseModel):
    inputs: ImportPCInputs = Field(description="Inputs for importing a point cloud")
    outputs: list[ImportPCOutputsCreate] = Field(description="List of outputs for point cloud import")


class ImportPCCost(BaseModel):
    mpoints: float = Field(description="Number of megapoints inside the point clouds", ge=0)


class Point3dTime(Point3d):
    t: float = Field(description="Timestamp of point")


class Scan(BaseModel):
    name: str = Field(description="Name of the scan")
    num_points: int = Field(description="Number of points in the scan", alias="numPoints")
    has_color: bool = Field(description="True if Scan has color information", alias="hasColor")
    has_intensity: bool = Field(description="True if Scan has intensity information", alias="hasIntensity")
    has_classification: bool = Field(description="True if Scan has classification information",
                                     alias="hasClassification")
    position: Optional[Point3d] = Field(description="Scanner position if scan is static", default=None)
    trajectories: Optional[list[list[Point3dTime]]] = Field(description="List of trajectories if scan was mobile",
                                                            default=None)


class PodMetadata(BaseModel):
    min_res: float = Field(description="Minimum resolution of the point cloud")
    max_res: float = Field(description="Maximum resolution of the point cloud")
    mean_res: float = Field(description="Mean resolution of the point cloud")
    med_res: float = Field(description="Median resolution of the point cloud")
    min_intensity: int = Field(description="Minimum intensity", ge=-32768, le=32767)
    max_intensity: int = Field(description="Maximum intensity", ge=-32768, le=32767)
    srs: str = Field(description="Spatial Reference System definition")
    bounding: BoundingBox = Field(description="Bounding box of the PointCloud")
    scans: list[Scan] = Field(description="List of scans")

File: typescript/packages/reality-capture/src/specifications/import_point_cloud.ts
import { z } from "zod";
import { BoundingBoxSchema, Point3dSchema } from "./geometry";

export const ImportPCInputsSchema = z.object({
  scene: z.string().describe("Reality data id of ContextScene to process"),
  crsData: z.string().regex(/^bkt:.+/).optional().describe("Path in the bucket for CRS data."),
});
export type ImportPCInputs = z.infer<typeof ImportPCInputsSchema>;

export const ImportPCOutputsSchema = z.object({
  scanCollection: z.string().describe("Output reality data id for scan collection"),
  scene: z.string().optional().describe("Output reality data id for context scene referencing scan collection")
});
export type ImportPCOutputs = z.infer<typeof ImportPCOutputsSchema>;

export const ImportPCSpecificationsSchema = z.object({
  inputs: ImportPCInputsSchema.describe("Inputs for importing a point cloud"),
  outputs: ImportPCOutputsSchema.describe("Outputs for point cloud import")
});
export type ImportPCSpecifications = z.infer<typeof ImportPCSpecificationsSchema>;

export enum ImportPCOutputsCreate {
  SCAN_COLLECTION = "scanCollection",
  SCENE = "scene",
}

export const ImportPCSpecificationsCreateSchema = z.object({
  inputs: ImportPCInputsSchema.describe("Inputs for importing a point cloud"),
  outputs: z.array(z.nativeEnum(ImportPCOutputsCreate)).describe("List of outputs for point cloud import")
});
export type ImportPCSpecificationsCreate = z.infer<typeof ImportPCSpecificationsCreateSchema>;

export const ImportPCCostSchema = z.object({
  mpoints: z.number().min(0).describe("Number of megapoints inside the point clouds")
});
export type ImportPCCost = z.infer<typeof ImportPCCostSchema>;

export const Point3dTimeSchema = Point3dSchema.extend({
  t: z.number().describe("Timestamp of point")
});
export type Point3dTime = z.infer<typeof Point3dTimeSchema>;

export const ScanSchema = z.object({
  name: z.string().describe("Name of the scan"),
  numPoints: z.number().int().describe("Number of points in the scan"),
  hasColor: z.boolean().describe("True if Scan has color information"),
  hasIntensity: z.boolean().describe("True if Scan has intensity information"),
  hasClassification: z.boolean().describe("True if Scan has classification information"),
  position: Point3dSchema.optional().describe("Scanner position if scan is static"),
  trajectories: z.array(z.array(Point3dTimeSchema)).optional().describe("List of trajectories if scan was mobile")
});
export type Scan = z.infer<typeof ScanSchema>;

export const PodMetadataSchema = z.object({
  minRes: z.number().describe("Minimum resolution of the point cloud"),
  maxRes: z.number().describe("Maximum resolution of the point cloud"),
  meanRes: z.number().describe("Mean resolution of the point cloud"),
  medRes: z.number().describe("Median resolution of the point cloud"),
  minIntensity: z.number().int().min(-32768).max(32767).describe("Minimum intensity"),
  maxIntensity: z.number().int().min(-32768).max(32767).describe("Maximum intensity"),
  crs: z.string().describe("Coordinate Reference System definition"),
  bounding: BoundingBoxSchema.describe("Bounding box of the PointCloud"),
  scans: z.array(ScanSchema).describe("List of scans")
});
export type PodMetadata = z.infer<typeof PodMetadataSchema>;